qualtran/_infra/data_types.py

```diff
@@ -521,46 +521,120 @@ class QFxp(QDType):
-    to represent the number. E.g. Using `(bitsize = 8, num_frac = 6, signed = False)`
-    then $\pi$ \approx 3.140625 = 11.001001, where the . represents the decimal place.
+    real number is signed we store negative values in two's complement form. The first
+    bit can therefore be treated as the sign bit in such cases (0 for +, 1 for -).
+    In total there are `bitsize = (num_int + num_frac)` bits used to represent the number.
+    E.g. Using `(bitsize = 8, num_frac = 6, signed = False)` then
+    $\pi$ \approx 3.140625 = 11.001001, where the . represents the decimal place.
 
     We can specify a fixed point real number by the tuple bitsize, num_frac and
-    signed, with num_int determined as `(bitsize - num_frac - n_sign)`.
+    signed, with num_int determined as `(bitsize - num_frac)`.
 
     Attributes:
         bitsize: The total number of qubits used to represent the integer and
             fractional part combined.
         num_frac: The number of qubits used to represent the fractional part of the real number.
-        signed: Whether the number is signed or not. If signed is true the
-            number of integer bits is reduced by 1.
+        signed: Whether the number is signed or not.
     """
 
     bitsize: SymbolicInt
     num_frac: SymbolicInt
     signed: bool = False
 
+    def __attrs_post_init__(self):
+        if not is_symbolic(self.num_qubits) and self.num_qubits == 1 and self.signed:
+            raise ValueError("num_qubits must be > 1.")
+        if not is_symbolic(self.bitsize) and not is_symbolic(self.num_frac):
+            if self.signed and self.bitsize == self.num_frac:
+                raise ValueError("num_frac must be less than bitsize if the QFxp is signed.")
+            if self.bitsize < self.num_frac:
+                raise ValueError("bitsize must be >= num_frac.")
+
     @property
     def num_qubits(self):
         return self.bitsize
 
     @property
     def num_int(self) -> SymbolicInt:
-        return self.bitsize - self.num_frac - int(self.signed)
+        return self.bitsize - self.num_frac
+
+    def is_symbolic(self) -> bool:
+        return is_symbolic(self.bitsize, self.num_frac)
 
     @property
-    def fxp_dtype_str(self) -> str:
-        return f'fxp-{"us"[self.signed]}{self.bitsize}/{self.num_frac}'
+    def _int_qdtype(self) -> Union[QUInt, QInt]:
+        """The corresponding integer type used to represent raw values of this type.
+
+        This raw integer value is used in the classical simulator to represent values
+        of QFxp registers.
+
+        For example, QFxp(6, 2) has 2 int bits and 4 frac bits, and the corresponding
+        int type is QUInt(6). So a true classical value of `10.0011` will have a raw
+        integer representation of `100011`.
+        """
+        return QInt(self.bitsize) if self.signed else QUInt(self.bitsize)
+
+    def get_classical_domain(self) -> Iterable[int]:
+        yield from self._int_qdtype.get_classical_domain()
```

Comment:
this needs a docstring, since it would be surprising to an unsuspecting user. If you bulk out the class docstring, it can be a one-liner that refers to the complete description in the class docstring.

Reply:
added